beater/java_attacher/java_attacher.go

```diff
@@ -0,0 +1,141 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package javaattacher
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/elastic/beats/v7/libbeat/logp"
+
+	"github.com/elastic/apm-server/beater/config"
+)
+
+type JavaAttacher struct {
+	cfg    config.JavaAttacherConfig
+	logger *logp.Logger
+}
+
+func New(cfg config.JavaAttacherConfig) (JavaAttacher, error) {
+	if cfg.JavaBin == "" {
+		if jh := os.Getenv("JAVA_HOME"); jh != "" {
+			cfg.JavaBin = filepath.Join(jh, "/bin/java")
+		} else {
+			bin, err := exec.LookPath("java")
+			if err != nil {
+				return JavaAttacher{}, fmt.Errorf("no java binary found: %v", err)
+			}
+			cfg.JavaBin = bin
+		}
+	} else {
+		// Ensure we're using the correct separators for the system
+		// running apm-server
+		cfg.JavaBin = filepath.FromSlash(cfg.JavaBin)
+	}
+	logger := logp.NewLogger("java-attacher")
+	return JavaAttacher{cfg: cfg, logger: logger}, nil
+}
+
+// javaAttacher is bundled by the server
+// TODO: Figure out the real path
+const javaAttacher = "/bin/apm-agent-attach-cli-1.24.0-slim.jar"
+
+func (j JavaAttacher) Run(ctx context.Context) error {
+	cmd := j.build(ctx)
+	j.logger.Debugf("starting java attacher with command: %s", strings.Join(cmd.Args, " "))
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+
+	// TODO: How do we want to handle logging?
+	// See comment:
+	// https://github.com/elastic/apm-server/issues/4830#issuecomment-863207642
+	donec := make(chan struct{})
+	defer close(donec)
+	go func() {
+		scanner := bufio.NewScanner(stdout)
+		b := struct {
+			LogLevel string `json:"log.level"`
+			Message  string `json:"message"`
+		}{}
+		for scanner.Scan() {
+			select {
+			case <-ctx.Done():
+				return
+			case <-donec:
+				// TODO: Is this necessary, or will the scanner
+				// finish when the command exits?
+				return
+			default:
+			}
+			json.Unmarshal(scanner.Bytes(), &b)
+			switch b.LogLevel {
+			case "FATAL", "ERROR":
+				j.logger.Error(b.Message)
+			case "WARN":
+				j.logger.Warn(b.Message)
+			case "INFO":
+				j.logger.Info(b.Message)
+			case "DEBUG", "TRACE":
+				j.logger.Debug(b.Message)
+			default:
+				j.logger.Errorf("unrecognized java-attacher log.level: %s", b.LogLevel)
+			}
+		}
+		if err := scanner.Err(); err != nil {
+			j.logger.Errorf("error scanning attacher logs: %v", err)
+		}
+	}()
+
+	return cmd.Wait()
+}
+
+func (j JavaAttacher) build(ctx context.Context) *exec.Cmd {
+	args := append([]string{"-jar", javaAttacher}, j.formatArgs()...)
+	return exec.CommandContext(ctx, j.cfg.JavaBin, args...)
+}
+
+func (j JavaAttacher) formatArgs() []string {
+	args := []string{"--continuous"}
```

Comment:
If you can query the log level of APM Server, you could set `--log-level` to one of `<off|fatal|error|warn|info|debug|trace|all>`.

Reply:
@axw is there an easy way to do this within the code?

Reply:
I'm not aware of a way to do that. It may be necessary to update libbeat/logp to expose the level.

Reply:
If the log level can't be directly read from the logger itself, is it possible to read the `logging.level` config option here?

Reply:
Perhaps. It's also possible to enable debug level for specific "selectors". For example we could introduce a "java-attacher" selector, and you could enable debug logging for just that by passing `-d java-attacher` to the apm-server command.

What we can do with the existing libbeat/logp API is to check if debug logging for the selector (or the process overall) is enabled: https://github.com/elastic/beats/blob/c3925fcb5435df6397803542492398da3e8ed484/libbeat/logp/logger.go#L111-L112

I think the pragmatic solution here is to check that and pass `--log-level=debug` if enabled, otherwise pass `--log-level=info`. We'll still parse the log records and call the appropriate logger method (Infof, Warnf, Errorf, etc.), so if the server is configured to only log warnings then the info logs will just get dropped after being parsed.

Reply:
It's possible to get the `logging.level` setting from the parent config, using the same code from sending the config to kibana:

```go
	cfg := ucfg.Config(*s.rawConfig)
	parentCfg := cfg.Parent()
```

And from there accessing `logging.level`. If it's empty, then we use the default `info` level.